apps/trading/components/swap/swap-form.tsx

```
@@ -0,0 +1,386 @@
+) => {
+  if (!marketData) return;
+  const isBid = side === Side.SIDE_SELL;
+  const price = isBid
+    ? new BigNumber(marketData.bestBidPrice).times(1 - toleranceFactor)
+    : new BigNumber(marketData.bestOfferPrice).times(1 + toleranceFactor);
+  return roundUpToTickSize(price, market.tickSize, isBid);
+};
+
+export const SwapForm = ({
+  marketData,
+  baseAsset,
+  quoteAsset,
+  market,
+  accounts,
+  assets: spotAssets,
+  setBaseAsset,
+  setQuoteAsset,
+}: {
+  market?: MarketFieldsFragment;
+  marketData: MarketDataFieldsFragment | null;
+  assets: AssetFieldsFragment[];
+  baseAsset?: AssetFieldsFragment;
+  quoteAsset?: AssetFieldsFragment;
+  accounts?: Account[] | null;
+  setBaseAsset: (asset?: AssetFieldsFragment) => void;
+  setQuoteAsset: (asset?: AssetFieldsFragment) => void;
+}) => {
+  const t = useT();
+
+  const [quoteAmount, setQuoteAmount] = useState('');
+  const [baseAmount, setBaseAmount] = useState('');
+  const [tolerance, setTolerance] = useState('');
+
+  const { pubKey, isReadOnly } = useVegaWallet();
+  const create = useVegaTransactionStore((state) => state.create);
+
+  const quoteAssetBalance = getQuoteAssetBalance(quoteAsset, accounts);
+  const baseAssetBalance = getQuoteAssetBalance(baseAsset, accounts);
+
+  const side = useSide({ market, baseAsset, quoteAsset });
+
+  const marketPrice = useMarketPrice({ marketData, side });
+  const prevMarketPrice = usePrevious(marketPrice);
+
+  const handleSwapAssets = () => {
+    const newBaseAsset = quoteAsset;
+    const newQuoteAsset = baseAsset;
+    setBaseAsset(newBaseAsset);
+    setQuoteAsset(newQuoteAsset);
+
+    setBaseAmount(quoteAmount);
+    setQuoteAmount(baseAmount);
+  };
+
+  const onSubmit = (e: FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    // Check users is connected
+    if (isReadOnly || !pubKey) return;
+    if (!Number(quoteAmount)) return;
+
+    if (!market) {
+      throw new Error('could not dervie spot market for swap');
```

This should also be caught earlier

---

I agree. Actually I don't think we need to throw an error here